src/type-layout.md

```diff
@@ -285,40 +286,165 @@ the default `enum` size and alignment for the target platform's C ABI.
+different behavior whether the enum has fields or no fields. It is an error
+for [zero-variant enumerations] to have a primitive representation. Combining
+two primitive representations together is an error.
+
+> Note: Primitive representations are primarily intended for Rust code that
+> wants to interoperate with the idioms of preexisting C/C++ codebases. If you
+> have control over both the Rust and C code, such as using C as FFI glue
+> between Rust and some third language, then you should use a
+> [primitive representation](#primitive-representation-of-enums-with-fields)
+> instead.
+
+#### Primitive Representation of Field-less Enums
+
+For [field-less enums], primitive representations set the size and alignment to
+be the same as the primitive type of the same name. For example, a field-less
+enum with a `u8` representation can only have discriminants between 0 and 255
+inclusive.
+
+#### Primitive Representation of Enums With Fields
 
-For [C-like enumerations], they set the size and alignment to be the same as the
-primitive type of the same name. For example, a C-like enumeration with a `u8`
-representation can only have discriminants between 0 and 255 inclusive.
+The representation of a primitive representation enum is a `repr(C)` union of
+`repr(C)` structs for each variant with a field. The first field of each struct
+in the union is the primitive representation version of the enum with all fields
+removed ("the tag") and the remaining fields are the fields of that variant.
+
+> Note: This representation is unchanged if the tag is given its own member in
+> the union, should that make manipulation more clear for you (although to
+> follow the C++ standard the tag member should be wrapped in a `struct`).
+
+```rust
+// This enum has the same layout as ...
+#[repr(u8)]
+enum MyEnum {
+    A(u32),
+    B(f32, u64),
+    C { x: u32, y: u8 },
+    D,
+ }
+
+// ... this union.
+#[repr(C)]
+union MyEnumRepr {
+    A: MyVariantA,
+    B: MyVariantB,
+    C: MyVariantC,
+    D: MyVariantD,
+}
 
-It is an error for [zero-variant enumerations] to have a primitive
-representation.
+// This is the discriminant enum.
+#[repr(u8)]
+#[derive(Copy, Clone)]
+enum MyEnumDiscriminant { A, B, C, D }
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct MyVariantA(MyEnumDiscriminant, u32);
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct MyVariantB(MyEnumDiscriminant, f32, u64);
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct MyVariantC { tag: MyEnumDiscriminant, x: u32, y: u8 }
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+struct MyVariantD(MyEnumDiscriminant);
+```
 
-For all other enumerations, the layout is unspecified.
+> Note: `union`s with non-`Copy` fields are unstable, see [55149].
 
-Likewise, combining two primitive representations together is unspecified.
+Combining the `repr(C)` and a primitive representation is only defined for enums
+with fields. The primitive representation modifies the `repr(C)` by changing the
```

Comment:
Yes, much better. :) Some wording suggestions:
```suggestion
For enums with fields, it is also possible to combining `repr(C)` and a primitive representation.
This modifies the `repr(C)` by changing the
```
Also it would be nice to make "the `repr(C)`" in the 2nd line a link to the relevant section.

And finally, is it possible to summarize the difference between this and "normal" primitive reps in a sentence? Something like "union-of-structs vs struct-of-tag-and-unions"?